spec/src/consensus.rs

```diff
@@ -340,82 +383,124 @@ impl Consensus {
-                block_reward,
-                remainder_reward,            // remainder_reward
-                header.hash().to_owned(),    // last_block_hash_in_previous_epoch
-                header.number() + 1,         // start
-                next_epoch_length,           // length
-                difficulty                   // difficulty,
+        let (next_epoch_length, bound) = if last_uncles_count == 0 {
+            (
+                cmp::min(self.max_epoch_length(), last_epoch_length * TAU),
+                true,
             )
         } else {
-            let next_epoch_length = self.max_epoch_length();
-            let difficulty = cmp::max(self.min_difficulty().clone(), last_difficulty >> 1);
-
-            let block_reward = Capacity::shannons(self.epoch_reward().as_u64() / next_epoch_length);
-            let remainder_reward =
-                Capacity::shannons(self.epoch_reward().as_u64() % next_epoch_length);
-            EpochExt::new(
-                last_epoch.number() + 1,    // number
-                block_reward,
-                remainder_reward,           // remainder_reward
-                header.hash().to_owned(),   // last_block_hash_in_previous_epoch
-                header.number() + 1,        // start
-                next_epoch_length,          // length
-                difficulty                  // difficulty,
-            )
+            // o_ideal * (1 + o_i ) * L_ideal * C_i,m
+            let numerator = orphan_rate_target
+                * (&last_orphan_rate + U256::one())
+                * &epoch_duration_target_u256
+                * &last_epoch_length_u256;
+            // o_i * (1 + o_ideal ) * L_i
+            let denominator =
+                &last_orphan_rate * (orphan_rate_target + U256::one()) * &last_epoch_duration;
+            let raw_next_epoch_length = u256_low_u64((numerator / denominator).into_u256());
+
+            self.bounding_epoch_length(raw_next_epoch_length, last_epoch_length)
+        };
+
+        // (3) Determining the Next Epoch’s Difficulty
+        let next_epoch_length_u256 = U256::from(next_epoch_length);
+        let diff_numerator = RationalU256::new(
+            &adjusted_last_epoch_hash_rate * epoch_duration_target,
+            U256::one(),
+        );
+        let next_epoch_diff = if bound {
+            if last_orphan_rate.is_zero() {
+                let denominator = U256::from(next_epoch_length);
+                (diff_numerator / denominator).into_u256()
+            } else {
+                let orphan_rate_estimation = RationalU256::one()
+                    / (((&last_orphan_rate + U256::one())
+                        * &epoch_duration_target_u256
+                        * &last_epoch_length_u256
+                        / (&last_orphan_rate * &last_epoch_duration * &next_epoch_length_u256))
+                        .saturating_sub_u256(U256::one()));
+
+                let denominator = if orphan_rate_estimation.is_zero() {
```

Comment:
Is it possible `orphan_rate_estimation` is negative or larger than `orphan_rate_target`?

Reply:
@nirenzang is not sure about this, so we use `saturating_sub` here.

Reply:
What if the `orphan_rate_estimation` is larger than `orphan_rate_target`?

Reply:
It is possible `orphan_rate_estimation` than `orphan_rate_target`, is there a problem?

Reply:
I can't find a problem here.